### ISSUE TITLE
Predict multiple postings

### ISSUE THREAD

**Issue by yagebu** [beancount/smart_importer#68]
Currently, `PredictPostings` will only predict a single posting and not a list of postings. It seems to me that the latter would be preferable, is there a reason why this is not done?

**Comment by johannesjh** [comment]
machine learning predicts a single posting, for which it leaves the amount empty (beancount will figure out the missing amount automatically).

but if machine learning were to predict multiple postings, there is no way for it to know what would be the right amounts for each of those postings (and beancount would not be able to figure that out either)

**Comment by yagebu** [comment]
> but if machine learning were to predict multiple postings, there is no way for it to know what would be the right amounts for each of those postings (and beancount would not be able to figure that out either)

In this case, some manual intervention is required anyway and having all the accounts there would make that way easier. (also, if the amount is an exact match, like it would likely happen for recurring transactions, we could actually fill in the amounts)

**Comment by johannesjh** [comment]
hence the feature request that fava should populate the dropdown list with accounts as suggested by machine learning

**Comment by yagebu** [comment]
Sure, but that would still be more complicated than having the postings filled in by smart_importer.

**Comment by johannesjh** [comment]
alright, but how many postings should be added, depending on what circumstances? e.g., 
1. a number that should be predicted by the machine learning model (it currently can't do that, but could probably be extended)
2. a configurable number that would be the same for all imported transactions? (I can't quite imagine the usecase of how this would be useful)

**Comment by yagebu** [comment]
In my transactions, there's usually groups of postings that occur together. There's simple transactions that only have two legs, say `Expenses:Food` and my credit card. For the ones that have more than two legs, the groups of postings are usually the same. Some examples (I've left out the bank account, for which I'll do the import):

- paycheck: legs for all kinds of taxes and insurances that are deducted
- buying stocks: commission and portfolio
- rent: splitting the monthly rent into several expense categories

So I think the model shouldn't really predict accounts for postings and then decide on some number, but rather it should classify by sets or tuples of accounts.

**Comment by yagebu** [comment]
The current PredictPostings is odd - if I understand it correctly, for every transaction with say n postings, it will add (n)(n-1)/2 TxnPostingAccount's namedtuples to the training data (one for every pair of postings). That seems to me like it would give much greater weight to these transactions...

If we instead used tuples as prediction targets, we could avoid the need for this additional step of modifying the training data and for two-posting transactions, it shouldn't make a difference.

**Comment by yagebu** [comment]
> What you are proposing can be implemented with multilabel classification in scikit-learn.

Actually, I think, I'm proposing something slightly differently: The classification targets would still be single labels, but instead of account names, it would be the tuples of account names. That would be preferable in my use case, as the account combinations will actually not be arbitrary combinations of relevant accounts but always the same fixed combination of accounts (and additionally, they would be ordered which I'd like).

> Note: I just found out on the same page that scikit-learn can do multiple classification tasks at once, called "multi-task classification". This would probably allow to predict postings and payees in a single step. That would simplify things a lot, we would not need two separate decorators anymore.

I think it's not too bad to have multiple decorators, chaining them in a certain order can give better results I think (cleaning up the payees first, and then using that to predict accounts for example). Having the option to do the prediction in one step can still be useful though.

I think with the changes in #71, we're not very far from only having to specify a minimal set of variables to implement a new "prediction decorator", so users could specify their own target attributes to predict or modify the set of weights.

**Comment by johannesjh** [comment]
Ok, so for training the machine learning pipeline, we would concatenate all account names that are involved in a transactions into a single string. The pipeline would likewise predict a single string of concatenated account names. I agree that this is a viable solution.

Regarding the ordering of account names: I think it would be beneficial to normalize the order of account names in the concatenated string prior to training. This would greatly reduce variance in training data. My assumption here is that it is unlikely for training data to be entirely consistent in how postings are ordered. Possible solutions for how this normalization could be achieved:
* One simple solution would be to sort the list of account names alphabetically prior to concatenating the string. This could be implemented as scikit-learn transformer, similar to the existing transformers in `machinelearning_helpers.py`. The drawback of this solution is that the pipeline would always predict alphabetically ordered account names, which would not be nice.
* A more refined solution would also normalize account names, but additionally learn the most popular ordering from training data, by noting the position of each account name in the list of postings. After prediction, the normalized string would be converted back into a list of account names, and the list would be sorted by average position This could also be implemented as scikit-learn transformer, but the transformer would need some additional state to keeps track of position data.

**Comment by yagebu** [comment]
This is now implemented in a basic way and replaces the old predictor (which had a multitude of inconsistencies).

> I think it would be beneficial to normalize the order of account names in the concatenated string prior to training. This would greatly reduce variance in training data. My assumption here is that it is unlikely for training data to be entirely consistent in how postings are ordered. 

Consistency and usefulness of the training data is really something that's up to the user.

I guess of all orders of accounts, we could just pick the most common. But even that currently doesn't seem worth the effort to me, I'd rather stick to the KISS principle here for now.
